src/client.rs

```diff
@@ -274,39 +317,68 @@ where
-                    if Arc::strong_count(&state.sending) == 1 {
-                        // `Sending` state with strong count of 1 can happen only
-                        // when the connection is dropped just before it reads our
-                        // event. In this case we treat is with the same way as `Poisoned`
-                        // state.
-                        true
-                    } else {
-                        // ClientConnectionHandler will wake us when we can retry
+            // Clear out bad connections. In case of a bad connection we
+            // must send the full wantlist because we don't know what
+            // the remote peer has received.
+            match state.sending_state {
+                SendingState::Ready => {
+                    // Allowed to send
+                }
+                SendingState::Requested(instant, connection_id) => {
+                    if instant.elapsed() < RECEIVE_REQUEST_TIMEOUT {
+                        // Sending in progress
                         continue;
                     }
+                    // Bad connection.
+                    // `ClientConnectionHandler` didn't receive `SendWantlist` request within time.
+                    state.established_connections.remove(&connection_id);
+                    state.send_full = true;
+                    state.sending_state = SendingState::Ready;
+                }
+                SendingState::RequestReceived(..) => {
+                    // Stream allocation in progress
+                    continue;
+                }
+                SendingState::Sending(..) => {
+                    // Sending in progress
+                    continue;
                 }
-                SendingState::Ready => state.send_full,
-                // State is poisoned, send full list to recover.
-                SendingState::Poisoned => true,
+                SendingState::Failed(connection_id) => {
+                    // Bad connection.
+                    // `ClientConnectionHandler` failed to send wantlist because of network issues.
+                    state.established_connections.remove(&connection_id);
+                    state.send_full = true;
+                    state.sending_state = SendingState::Ready;
+                }
+            };
+
+            let Some(connection_id) = state.established_connections.iter().next().copied() else {
+                peers_without_connection.push(*peer);
+                continue;
             };
 
-            let wantlist = if send_full {
+            let wantlist = if state.send_full {
                 state.wantlist.generate_proto_full(&self.wantlist)
             } else {
+                // NOTE: `generate_proto_update` alters the internal state of `WantlistState`
+                // each time it is called. So after calling it, any error should be recovered
+                // with `send_full`, even if the error happens before any byte leaves the wire.
```

Comment:
```suggestion
                // each time it is called. Receiving error when calling it requires performing recovery
                // with `send_full`, even if the error happens before any byte leaves the
```

Reply:
Could we potentially avoid some re-sends here if we could checkpoint the wantlist state and have the ability to roll-back last update (or few)? For future reference and issue creation.

Reply:
We can have a commit mechanism or something similar. But I'm not really sure if it worth the complexity.

Reply:
> Receiving error when calling it

It is not about receiving error when calling `generate_proto_update`. It is about receiving error after this point. For example when connection manager couldn't allocate a stream within time. Check this: https://github.com/eigerco/beetswap/pull/36#discussion_r1553555522